tests/evm/test_create.py

```diff
@@ -0,0 +1,371 @@
+            is_return,
+            create_contexts,
+            stack,
+            is_warm_access,
+            expected(
+                opcode,
+                create_contexts,
+                stack,
+            ),
+        )
+        for opcode, is_return, create_contexts, stack, is_warm_access in product(
+            opcodes, is_return, create_contexts, stacks, is_warm_accesss
+        )
+    ]
+
+
+TESTING_DATA = gen_testing_data()
+
+
+@pytest.mark.parametrize(
+    "opcode, caller, is_return, caller_ctx, stack, is_warm_access, expected",
+    TESTING_DATA,
+)
+def test_create_create2(
+    opcode: Opcode,
+    caller: Account,
+    is_return: Bytecode,
+    caller_ctx: CreateContext,
+    stack: Stack,
+    is_warm_access: bool,
+    expected: Expected,
+):
+    randomness = rand_fq()
+    CURRENT_CALL_ID = 1
+
+    init_codes = gen_bytecode(is_return, stack.offset)
+
+    is_create2 = 1 if opcode == Opcode.CREATE2 else 0
+    if is_create2 == 1:
+        caller_bytecode = init_codes.create2(
+            stack.value,
+            stack.offset,
+            stack.size,
+            stack.salt,
+        ).stop()
+    else:  # CREATE
+        caller_bytecode = init_codes.create(
+            stack.value,
+            stack.offset,
+            stack.size,
+        ).stop()
+
+    init_bytecode = init_codes
+    init_bytecode_hash = RLC(init_bytecode.hash(), randomness)
+
+    nonce = caller.nonce
+    is_success = is_return
+    is_reverted_by_caller = not caller_ctx.is_persistent and is_success
+    is_reverted_by_callee = not is_success
+    callee_is_persistent = caller_ctx.is_persistent and is_success
+    callee_rw_counter_end_of_reversion = (
+        80
+        if is_reverted_by_callee
+        else (
+            caller_ctx.rw_counter_end_of_reversion - (caller_ctx.reversible_write_counter + 1)
+            if is_reverted_by_caller
+            else 0
+        )
+    )
+
+    if is_create2 == 1:
+        preimage = (
+            b"\xff"
+            + caller.address.to_bytes(20, "big")
+            + int(stack.salt).to_bytes(32, "little")
+            + init_bytecode_hash.le_bytes
+        )
+        contract_addr = keccak256(preimage)
+    else:
+        contract_addr = keccak256(rlp.encode([caller.address.to_bytes(20, "big"), caller.nonce]))
+    contract_address = int.from_bytes(contract_addr[-20:], "big")
+
+    # can't be a static all
+    is_static = 0
+
+    next_call_id = 65
+    rw_counter = next_call_id
+    # CREATE: 33 * 3(push) + 1(CREATE) + 1(mstore) + 33(PUSH32) + 2(PUSH) + 1(RETURN)
+    # CREATE2: 33 * 4(push) + 1(CREATE) + 1(mstore) + 33(PUSH32) + 2(PUSH) + 1(RETURN)
+    next_program_counter = 33 * 4 + 1 + 35 + 1 if is_create2 else 33 * 3 + 1 + 35 + 1
+    assert caller_bytecode.code[next_program_counter] == opcode
+
+    # CREATE: 1024 - 3 + 1 = 1022
+    # CREATE2: 1024 - 4 + 1 = 1021
+    stack_pointer = 1021 - is_create2
+
+    # first element of init_bytecode is length of code, so skip it
```

Comment:
hmm is this the reason the `dst_addr` starts with `1`, but is this rule in EVM spec?

Reply:
fixed in [8c49721](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/355/commits/8c4972148e3ca79d57e50711498be2170590fc59)